File: TinyPG/Examples/TestCpp/TinyExe/src/StaticFunction.cpp
#include "StaticFunction.h"

namespace TinyExe
{
	StaticFunction::StaticFunction(const std::string& name, FunctionDelegate function, int minParameters, int maxParameters)
	{
		Name = name;
		functionDelegate = function;
		MinParameters = minParameters;
		MaxParameters = maxParameters;
		Arguments = new Variables();
	}

	StaticFunction::StaticFunction(const std::string& name, FunctionContextDelegate function, int minParameters, int maxParameters)
	{
		Name = name;
		functionContextDelegate = function;
		MinParameters = minParameters;
		MaxParameters = maxParameters;
		Arguments = new Variables();
	}

	std::any StaticFunction::Eval(std::vector<std::any> parameters, ParseTree* tree)
	{
		tree->Context->PushScope(NULL);

		std::any result;
		if (functionDelegate != NULL)
			result = functionDelegate(parameters);
		else if (functionContextDelegate != NULL)
			result = functionContextDelegate(parameters, tree->Context);
		tree->Context->PopScope();
		return result;
	}

}

File: TinyPG/Examples/TestJava/TinyExe/src/tinyexe/StaticFunction.java
package tinyexe;

public class StaticFunction extends Function
{
    private FunctionDelegate functionDelegate;
    private FunctionContextDelegate functionContextDelegate;
    /// <summary>
    /// the actual function implementation
    /// </summary>
    public FunctionDelegate getFunctionDelegate() { return functionDelegate; }
    public void setFunctionDelegate(FunctionDelegate val) { functionDelegate = val; }

    public FunctionContextDelegate getFunctionContextDelegate() { return functionContextDelegate; }
    public void setFunctionContextDelegate(FunctionContextDelegate val) { functionContextDelegate = val; }

    @Override
    public Object Eval(Object[] parameters, ParseTree tree)
    {
        tree.Context.PushScope(null);

        Object result = null;
        if (functionDelegate != null)
            result = functionDelegate.invoke(parameters);
        else if (functionContextDelegate != null)
            result = functionContextDelegate.invoke(parameters, tree.Context);
        tree.Context.PopScope();
        return result;
    }

    public StaticFunction(String name, FunctionDelegate function, int minParameters, int maxParameters)
    {
        this.name=name;
        functionDelegate = function;
        this.minParameters = minParameters;
        this.maxParameters = maxParameters;
        this.arguments = new Variables();
    }

    public StaticFunction(String name, FunctionContextDelegate function, int minParameters, int maxParameters)
    {
        this.name = name;
        functionContextDelegate = function;
        this.minParameters = minParameters;
        this.maxParameters = maxParameters;
        this.arguments = new Variables();
    }
}